langchain/chains/query_constructor/query_language.py

```diff
@@ -0,0 +1,72 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Any, List, Optional, Union
+
+from pydantic import BaseModel
+
+
+class Visitor(ABC):
+    """Abstract visitor interface."""
+
+    allowed_comparators: Optional[List["Comparator"]] = None
+    allowed_operators: Optional[List["Operator"]] = None
+
+    @abstractmethod
+    def visit_operation(self, operation: "Operation") -> Any:
+        """"""
+
+    @abstractmethod
+    def visit_comparison(self, comparison: "Comparison") -> Any:
+        """"""
+
+    @abstractmethod
+    def visit_structured_query(self, structured_query: "StructuredQuery") -> Any:
+        """"""
+
+
+def _to_snake_case(name: str) -> str:
+    """Convert a name into snake_case."""
+    snake_case = ""
+    for i, char in enumerate(name):
+        if char.isupper() and i != 0:
+            snake_case += "_" + char.lower()
+        else:
+            snake_case += char.lower()
+    return snake_case
+
+
+class Expr(BaseModel):
+    def accept(self, visitor: Visitor) -> Any:
+        return getattr(visitor, f"visit_{_to_snake_case(self.__class__.__name__)}")(
+            self
+        )
+
+
+class Operator(str, Enum):
+    AND = "and"
+    OR = "or"
+    NOT = "not"
+
+
+class Comparator(str, Enum):
+    EQ = "eq"
```

Comment:
Do we want `NEQ` or does or does it make sense to only rely on NOT + EQ so there's a unique representation?

----------------

One edge case (there will probably be others): 

"documents that have a publication date" (we don't have a primitive type for `null` right now in the grammar)